File: plugins/CCreator/src/ccreator/spice2schematic/router.py
"""Manhattan wire router for ngspice -> schematic conversion.

Pin offsets (relative to instance x,y):
  res/capa/ind/diode/vsource/isource:  p=(0,-30)  n=(0,+30)
  nmos4:  d=(+20,-30)  g=(-20,0)  s=(+20,+30)  b=(+20,0)
  pmos4:  d=(+20,+30)  g=(-20,0)  s=(+20,-30)  b=(+20,0)
  npn:    c=(+20,-30)  b=(-20,0)  e=(+20,+30)
  pnp:    c=(+20,+30)  b=(-20,0)  e=(+20,-30)
  njfet:  d=(+20,-30)  g=(-20,0)  s=(+20,+30)
  pjfet:  d=(+20,+30)  g=(-20,0)  s=(+20,-30)
  vdd:    vdd=(0,+10)
  gnd:    gnd=(0,-10)
"""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass
from enum import Enum
from typing import Optional

from .layout import PlacedElement, is_gnd_net, is_vdd_net
from .parser import Element


@dataclass
class RouteWire:
    x0: int
    y0: int
    x1: int
    y1: int
    net_name: Optional[str] = None


class PowerKind(Enum):
    VDD = "vdd"
    GND = "gnd"


@dataclass
class PowerSym:
    kind: PowerKind
    x: int
    y: int


@dataclass
class RouteResult:
    wires: list[RouteWire]
    power: list[PowerSym]


@dataclass
class ConnEntry:
    pin: str
    net: str


# Pin offset tables
@dataclass
class PinOff:
    name: str
    dx: int
    dy: int


TWO_TERM = [PinOff("p", 0, -30), PinOff("n", 0, 30)]

NMOS_PINS = [
    PinOff("d", 20, -30),
    PinOff("g", -20, 0),
    PinOff("s", 20, 30),
    PinOff("b", 20, 0),
]

PMOS_PINS = [
    PinOff("d", 20, 30),
    PinOff("g", -20, 0),
    PinOff("s", 20, -30),
    PinOff("b", 20, 0),
]

NPN_PINS = [PinOff("c", 20, -30), PinOff("b", -20, 0), PinOff("e", 20, 30)]

PNP_PINS = [PinOff("c", 20, 30), PinOff("b", -20, 0), PinOff("e", 20, -30)]

JFET_N_PINS = [PinOff("d", 20, -30), PinOff("g", -20, 0), PinOff("s", 20, 30)]

JFET_P_PINS = [PinOff("d", 20, 30), PinOff("g", -20, 0), PinOff("s", 20, -30)]


def _pin_offsets(sym_name: str) -> list[PinOff]:
    table = {
        "nmos4": NMOS_PINS,
        "pmos4": PMOS_PINS,
        "npn": NPN_PINS,
        "pnp": PNP_PINS,
        "njfet": JFET_N_PINS,
        "pjfet": JFET_P_PINS,
    }
    return table.get(sym_name, TWO_TERM)


def pin_pos(p: PlacedElement, node_idx: int) -> Optional[tuple[int, int]]:
    """Absolute pin position from placed element + node index."""
    offs = _pin_offsets(p.sym.sym)
    if node_idx >= len(offs):
        return None
    return (p.x + offs[node_idx].dx, p.y + offs[node_idx].dy)


def build_conns(elem: Element, sym_name: str) -> list[ConnEntry]:
    """Build conn list mapping symbol pin names to net names."""
    offs = _pin_offsets(sym_name)
    count = min(len(elem.nodes), len(offs))
    return [ConnEntry(pin=offs[i].name, net=elem.nodes[i]) for i in range(count)]


def route(
    elements: list[Element], placed: list[PlacedElement]
) -> RouteResult:
    """Route all nets from placed element list."""
    # Build net -> [(elem_idx, node_idx)] map
    net_pins: dict[str, list[tuple[int, int]]] = defaultdict(list)
    for ei, elem in enumerate(elements):
        for ni, node in enumerate(elem.nodes):
            net_pins[node].append((ei, ni))

    wires: list[RouteWire] = []
    power: list[PowerSym] = []

    for net, pins in net_pins.items():
        if is_gnd_net(net):
            for ei, ni in pins:
                pos = pin_pos(placed[ei], ni)
                if pos:
                    power.append(PowerSym(kind=PowerKind.GND, x=pos[0], y=pos[1] + 10))
            continue

        if is_vdd_net(net):
            for ei, ni in pins:
                pos = pin_pos(placed[ei], ni)
                if pos:
                    power.append(PowerSym(kind=PowerKind.VDD, x=pos[0], y=pos[1] - 10))
            continue

        # Collect pin positions for this net
        pts: list[tuple[int, int]] = []
        for ei, ni in pins:
            pos = pin_pos(placed[ei], ni)
            if pos:
                pts.append(pos)

        if len(pts) < 2:
            continue

        for i in range(1, len(pts)):
            _route_segment(wires, pts[i - 1], pts[i], net)

    return RouteResult(wires=wires, power=power)


def _route_segment(
    wires: list[RouteWire],
    p1: tuple[int, int],
    p2: tuple[int, int],
    net: str,
) -> None:
    """Emit one L-shaped wire from p1 to p2 (horizontal first, then vertical)."""
    if p1 == p2:
        return

    if p1[0] == p2[0] or p1[1] == p2[1]:
        wires.append(RouteWire(x0=p1[0], y0=p1[1], x1=p2[0], y1=p2[1], net_name=net))
        return

    # L-shape: horizontal then vertical
    elbow = (p2[0], p1[1])
    wires.append(RouteWire(x0=p1[0], y0=p1[1], x1=elbow[0], y1=elbow[1], net_name=net))
    wires.append(RouteWire(x0=elbow[0], y0=elbow[1], x1=p2[0], y1=p2[1], net_name=net))

File: .src/external/spice2schematic/router.zig
// router.zig — Manhattan wire router for SPICE-to-schematic conversion.
//
// Connects pins by net name using orthogonal (Manhattan) routing.
// For each net with 2+ pins, creates L-shaped wire segments between
// consecutive pin positions. Power/ground nets get power symbols instead.

const std = @import("std");
const Allocator = std.mem.Allocator;
const List = std.ArrayListUnmanaged;
const parser = @import("parser.zig");
const layout = @import("layout.zig");
const core = @import("core");
const DeviceKind = core.types.DeviceKind;

// ── Public types ────────────────────────────────────────────────────────────

pub const RouteWire = struct {
    x0: i32,
    y0: i32,
    x1: i32,
    y1: i32,
    net_name: []const u8,
};

pub const PowerKind = enum(u1) {
    vdd,
    gnd,
};

pub const PowerSym = struct {
    kind: PowerKind,
    x: i32,
    y: i32,
};

pub const RouteResult = struct {
    wires: []const RouteWire,
    power: []const PowerSym,
};

// ── Pin offset tables ───────────────────────────────────────────────────────

const PinOff = struct {
    dx: i32,
    dy: i32,
};

const TWO_TERM = [_]PinOff{
    .{ .dx = 0, .dy = -30 }, // p (positive)
    .{ .dx = 0, .dy = 30 }, // n (negative)
};

const NMOS_PINS = [_]PinOff{
    .{ .dx = 20, .dy = -30 }, // d (drain)
    .{ .dx = -20, .dy = 0 }, // g (gate)
    .{ .dx = 20, .dy = 30 }, // s (source)
    .{ .dx = 20, .dy = 0 }, // b (bulk)
};

const PMOS_PINS = [_]PinOff{
    .{ .dx = 20, .dy = 30 }, // d (drain)
    .{ .dx = -20, .dy = 0 }, // g (gate)
    .{ .dx = 20, .dy = -30 }, // s (source)
    .{ .dx = 20, .dy = 0 }, // b (bulk)
};

const NPN_PINS = [_]PinOff{
    .{ .dx = 20, .dy = -30 }, // c (collector)
    .{ .dx = -20, .dy = 0 }, // b (base)
    .{ .dx = 20, .dy = 30 }, // e (emitter)
};

const PNP_PINS = [_]PinOff{
    .{ .dx = 20, .dy = 30 }, // c (collector)
    .{ .dx = -20, .dy = 0 }, // b (base)
    .{ .dx = 20, .dy = -30 }, // e (emitter)
};

const JFET_N_PINS = [_]PinOff{
    .{ .dx = 20, .dy = -30 }, // d
    .{ .dx = -20, .dy = 0 }, // g
    .{ .dx = 20, .dy = 30 }, // s
};

const JFET_P_PINS = [_]PinOff{
    .{ .dx = 20, .dy = 30 }, // d
    .{ .dx = -20, .dy = 0 }, // g
    .{ .dx = 20, .dy = -30 }, // s
};

const FOUR_TERM_CTRL = [_]PinOff{
    .{ .dx = 0, .dy = -30 }, // n+
    .{ .dx = 0, .dy = 30 }, // n-
    .{ .dx = -30, .dy = -30 }, // nc+
    .{ .dx = -30, .dy = 30 }, // nc-
};

fn pinOffsets(kind: DeviceKind) []const PinOff {
    return switch (kind) {
        .nmos4 => &NMOS_PINS,
        .pmos4 => &PMOS_PINS,
        .npn => &NPN_PINS,
        .pnp => &PNP_PINS,
        .njfet => &JFET_N_PINS,
        .pjfet => &JFET_P_PINS,
        .vcvs, .vccs => &FOUR_TERM_CTRL,
        else => &TWO_TERM,
    };
}

pub fn pinPos(placed: layout.PlacedElement, node_idx: usize) ?struct { x: i32, y: i32 } {
    const offs = pinOffsets(placed.kind);
    if (node_idx >= offs.len) return null;
    return .{
        .x = placed.x + offs[node_idx].dx,
        .y = placed.y + offs[node_idx].dy,
    };
}

// ── Routing algorithm ───────────────────────────────────────────────────────

const PinRef = struct {
    elem_idx: u32,
    node_idx: u32,
};

const Point = struct { x: i32, y: i32 };

/// Route all nets from placed element list.
pub fn route(
    arena: Allocator,
    elements: []const parser.Element,
    placed: []const layout.PlacedElement,
) !RouteResult {
    if (elements.len == 0 or placed.len == 0) {
        return .{ .wires = &.{}, .power = &.{} };
    }

    // Build net -> [(elem_idx, node_idx)] map
    var net_pins = std.StringHashMapUnmanaged(List(PinRef)){};

    for (elements, 0..) |elem, ei| {
        for (elem.nodes, 0..) |node, ni| {
            const gop = try net_pins.getOrPut(arena, node);
            if (!gop.found_existing) {
                gop.value_ptr.* = .{};
            }
            try gop.value_ptr.append(arena, .{
                .elem_idx = @intCast(ei),
                .node_idx = @intCast(ni),
            });
        }
    }

    var wires: List(RouteWire) = .{};
    var power: List(PowerSym) = .{};

    var iter = net_pins.iterator();
    while (iter.next()) |entry| {
        const net = entry.key_ptr.*;
        const pins = entry.value_ptr.items;

        // Ground nets -> place GND symbols
        if (layout.isGndNet(net)) {
            for (pins) |pin| {
                if (pin.elem_idx < placed.len) {
                    if (pinPos(placed[pin.elem_idx], pin.node_idx)) |pos| {
                        try power.append(arena, .{ .kind = .gnd, .x = pos.x, .y = pos.y + 10 });
                    }
                }
            }
            continue;
        }

        // VDD nets -> place VDD symbols
        if (layout.isVddNet(net)) {
            for (pins) |pin| {
                if (pin.elem_idx < placed.len) {
                    if (pinPos(placed[pin.elem_idx], pin.node_idx)) |pos| {
                        try power.append(arena, .{ .kind = .vdd, .x = pos.x, .y = pos.y - 10 });
                    }
                }
            }
            continue;
        }

        // Regular net: collect pin positions, route pairwise
        var pts: List(Point) = .{};
        for (pins) |pin| {
            if (pin.elem_idx < placed.len) {
                if (pinPos(placed[pin.elem_idx], pin.node_idx)) |pos| {
                    try pts.append(arena, .{ .x = pos.x, .y = pos.y });
                }
            }
        }

        if (pts.items.len < 2) continue;

        // Chain route: connect consecutive pin positions
        for (1..pts.items.len) |i| {
            try routeSegment(arena, &wires, pts.items[i - 1], pts.items[i], net);
        }
    }

    return .{
        .wires = wires.items,
        .power = power.items,
    };
}

/// Emit one L-shaped Manhattan wire from p1 to p2.
fn routeSegment(
    arena: Allocator,
    wires: *List(RouteWire),
    p1: Point,
    p2: Point,
    net: []const u8,
) !void {
    if (p1.x == p2.x and p1.y == p2.y) return;

    // Straight line
    if (p1.x == p2.x or p1.y == p2.y) {
        try wires.append(arena, .{
            .x0 = p1.x,
            .y0 = p1.y,
            .x1 = p2.x,
            .y1 = p2.y,
            .net_name = net,
        });
        return;
    }

    // L-shape: horizontal then vertical
    const elbow_x = p2.x;
    const elbow_y = p1.y;
    try wires.append(arena, .{
        .x0 = p1.x,
        .y0 = p1.y,
        .x1 = elbow_x,
        .y1 = elbow_y,
        .net_name = net,
    });
    try wires.append(arena, .{
        .x0 = elbow_x,
        .y0 = elbow_y,
        .x1 = p2.x,
        .y1 = p2.y,
        .net_name = net,
    });
}

// ── Tests ───────────────────────────────────────────────────────────────────

test "pinPos — two terminal" {
    const placed = layout.PlacedElement{
        .elem_idx = 0,
        .x = 100,
        .y = -100,
        .kind = .resistor,
        .symbol = "res",
    };
    const p0 = pinPos(placed, 0).?;
    try std.testing.expectEqual(@as(i32, 100), p0.x);
    try std.testing.expectEqual(@as(i32, -130), p0.y);

    const p1 = pinPos(placed, 1).?;
    try std.testing.expectEqual(@as(i32, 100), p1.x);
    try std.testing.expectEqual(@as(i32, -70), p1.y);
}

test "pinPos — nmos4" {
    const placed = layout.PlacedElement{
        .elem_idx = 0,
        .x = 200,
        .y = 0,
        .kind = .nmos4,
        .symbol = "nmos4",
    };
    const d = pinPos(placed, 0).?;
    try std.testing.expectEqual(@as(i32, 220), d.x);
    try std.testing.expectEqual(@as(i32, -30), d.y);
    const g = pinPos(placed, 1).?;
    try std.testing.expectEqual(@as(i32, 180), g.x);
    try std.testing.expectEqual(@as(i32, 0), g.y);
}

test "pinPos — out of range" {
    const placed = layout.PlacedElement{
        .elem_idx = 0,
        .x = 0,
        .y = 0,
        .kind = .resistor,
        .symbol = "res",
    };
    try std.testing.expect(pinPos(placed, 5) == null);
}

test "route empty" {
    var arena_state = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena_state.deinit();
    const arena = arena_state.allocator();

    const result = try route(arena, &.{}, &.{});
    try std.testing.expectEqual(@as(usize, 0), result.wires.len);
    try std.testing.expectEqual(@as(usize, 0), result.power.len);
}
